output/mistral_fluentbit/mistral_fluentbit.c

```diff
@@ -0,0 +1,553 @@
+                *q = 'r';
+                break;
+            case '\t':
+                *q++ = '\\';
+                *q = 't';
+                break;
+            default:
+                *q = *p;
+                break;
+            }
+        }
+        /* Memory was allocated with calloc so string is already null terminated */
+        char *small_escaped = realloc(escaped, q - escaped + 1);
+        if (small_escaped) {
+            return small_escaped;
+        } else {
+            return escaped;
+        }
+    } else {
+        return NULL;
+    }
+}
+
+/*
+ * mistral_startup
+ *
+ * Required function that initialises the type of plug-in we are running. This
+ * function is called immediately on plug-in start-up.
+ *
+ * In addition this plug-in needs to initialise the Fluent Bit connection
+ * parameters. The stream used for error messages defaults to stderr but can
+ * be overridden here by setting plugin->error_log.
+ *
+ * Parameters:
+ *   plugin - A pointer to the plug-in information structure. This function
+ *            must set plugin->type before returning, if it doesn't the plug-in
+ *            will immediately shut down.
+ *   argc   - The number of entries in the argv array
+ *   argv   - A pointer to the argument array passed to main.
+ *
+ * Returns:
+ *   void - but see note about setting plugin->type above.
+ */
+void mistral_startup(mistral_plugin *plugin, int argc, char *argv[])
+{
+    /* Returning without setting plug-in type will cause a clean exit */
+
+    static const struct option options[] = {
+        {"error", required_argument, NULL, 'e'},
+        {"host", required_argument, NULL, 'h'},
+        {"port", required_argument, NULL, 'p'},
+        {0, 0, 0, 0},
+    };
+
+    const char *error_file = NULL;
+    const char *host = "localhost";
+    uint16_t port = 5170; /* This is the default TCP server port for Fluent Bit TCP connectivity */
+    int opt;
+    mode_t new_mode = 0;
+
+    /* This is thread safe because the receiving thread hasn't started at this point */
+    gettimeofday(&mistral_plugin_start, NULL);
+    timerclear(&mistral_plugin_end);
+
+    mistral_fluentbit_init(&fluentbit_tcp_ctx);
+
+    while ((opt = getopt_long(argc, argv, "e:h:p:v:", options, NULL)) != -1) {
+        switch (opt) {
+        case 'e':
+            error_file = optarg;
+            break;
+        case 'h':
+            host = optarg;
+            break;
+        case 'p': {
+            char *end = NULL;
+            unsigned long tmp_port = strtoul(optarg, &end, 10);
+            if (tmp_port == 0 || tmp_port > UINT16_MAX || !end || *end) {
+                mistral_err("Invalid port specified %s\n", optarg);
+                return;
+            }
+            port = (uint16_t)tmp_port;
+            break;
+        }
+        default:
+            usage(argv[0]);
+            return;
+        }
+    }
+
+    log_file_ptr = &(plugin->error_log);
+
+    if (error_file != NULL) {
+        if (new_mode > 0) {
```

Comment:
It looks as if `new_mode` is always zero, so this block never used?

Reply:
I added -m option back which sets the error log file properties. `new_mode` is used now.